src/main.rs

```diff
@@ -24,72 +20,45 @@ where
-            .map(move |_| println!("{}: Finished.", desc)),
+        f.map_err(move |e| eprintln!("{}: '{:?}'", desc, e))
+            .map(move |_| eprintln!("{}: Finished.", desc)),
     );
 }
 
 fn main() {
     let server = Server::bind("127.0.0.1:2794", &ReactorHandle::default()).unwrap();
 
-    let ToxHandle { request_tx, answer_rx } = spawn_tox();
-
-    let tox_tx = Arc::new(Mutex::new(request_tx));
-
-    let connection_sink = Arc::new(Mutex::new(None));
-    let connection_sink2 = connection_sink.clone();
-
-    let p = answer_rx
-        .map_err(|_| IoError::new(IoErrorKind::Other, "answer_rx dropped"))
-        .for_each(move |r| {
-            if let Some(ref mut sink) = *connection_sink2.lock().unwrap() {
-                let sink : &mut tokio::sync::mpsc::UnboundedSender<websocket::OwnedMessage> = sink;
-                let answer = serde_json::to_string(&r).unwrap();
-                sink.try_send(websocket::OwnedMessage::Text(answer))
-                    .map_err(|_|  IoError::new(IoErrorKind::Other, "connection_sink dropped"))?;
-            }
-
-            Ok(())
-        });
-
     let f = server
         .incoming()
         .then(future::ok) // wrap good and bad events into future::ok
         .filter(|event| {
             match event {
                 Ok(_) => true, // a good connection
                 Err(InvalidConnection { ref error, .. }) => {
-                    println!("Bad client: {}", error);
+                    eprintln!("Bad client: {}", error);
                     false // we want to save the stream if a client cannot make a valid handshake
                 }
             }
         })
         .and_then(|event| event) // unwrap good connections
         .map_err(|_| IoError::new(IoErrorKind::Other, "invalid connection"))
         .for_each(move |(upgrade, addr)| {
-            let connection_sink = connection_sink.clone();
-            let connection_sink2 = connection_sink.clone();
-            let tox_tx = tox_tx.clone();
-
-            println!("Got a connection from: {}", addr);
-            if connection_sink.lock().unwrap().is_some() {
-                spawn_future(upgrade.reject(), "Reject the second connection");
-                return Ok(());
-            }
+            eprintln!("Got a connection from: {}", addr);
+
+            let secret_key = upgrade.uri()
+                .get(1..)
+                .and_then(|sk| sk.parse().ok());
+
+            let ToxHandle { request_tx, answer_rx, guard } = spawn_tox(secret_key);
 
-            // accept the request to be a ws connection if it does
             let f = upgrade
                 .accept()
                 .map_err(|e| IoError::new(IoErrorKind::Other,
                     format!("websocket accept err: {}", e)
                 ))
                 .and_then(move |(s, _h)| {
                     let (sink, stream) = s.split();
-                    let tox_tx = (*tox_tx.lock().unwrap()).clone();
```

Comment:
You have to send SK back to client.

Reply:
Shouldn't a client get everything with `Info` request?

Reply:
Oh wait, it doesn't make much sense to call `Info` on a blank profile. But still sending SK could be done in an another PR (it requires a new event in ws-tox-protocol).